### ISSUE TITLE
Global variables undefined in interactive use of embedded ipython shell

### ISSUE THREAD

**Issue by ipython** [ipython/ipython#62]
Original Launchpad bug 399627: https://bugs.launchpad.net/ipython/+bug/399627
Reported by: h-fangohr (Hans Fangohr).

The error can be reproduced as follows:

1. Start Python, and start embedded ipython session. Following ipython's manual, we do

```
fangohr@eta:~$ python
Python 2.4.3 (#1, Jun  8 2009, 14:09:06) 
[GCC 4.1.2 20061115 (prerelease) (Debian 4.1.1-21)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from IPython.Shell import IPShellEmbed
>>> ipshell=IPShellEmbed()
>>> ipshell()
```

Within the just started ipython session, global variables are sometimes not visible. Two examples are:

Example 1:

```
In [1]: a=1

In [2]: def f(x):
   ...:     return a*x
   ...: 

In [3]: f(2)
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)

/home/fangohr/<ipython console> 

/home/fangohr/<ipython console> in f(x)

NameError: global name 'a' is not defined
```

Example 2:

```
In [4]: b=1

In [5]: (lambda :b)()
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)

/home/fangohr/<ipython console> 

/home/fangohr/<ipython console> in <lambda>()

NameError: global name 'b' is not defined
```

There is no error if "b=1; (lambda :b)()" is put in a script, and this script is executed using ipython's "run" command.

There is no error if "b=1; (lambda :b)()" is executed interactively after having started ipython.

The only way the error seems comes up is if an embedded IPython shell is started from a Python prompt AND the commands are executed interactively at the prompt.

I find the same error when trying ipython-0.9.1 (with python2.4).

The bug was reported by Olivier Klein to the nmag team (http://nmag.soton.ac.uk).

**Comment by philippschw** [comment]
Problem still persists in ipython==4.2.0. Interestingly under windows it is no problem but together with ubuntu global variables are not recognized.

Use case:
```
from ipywidgets import interact, FloatSlider, IntSlider,RadioButtons, Dropdown

@interact(sv1 = radio_1, Scenario = drop_1, sv2 = slider_2)
def update_map(sv1,sv2, Scenario):
```

**Comment by blackball** [comment]
Problem remains. IPython 3.1.0, Debian Whezzy.

**Comment by anfedorov** [comment]
This was fixed for a little while, seems to be reintroduced (IPython 5.1.0 on macOS)

**Comment by Erotemic** [comment]
Just in case this helps anyone I've been using this piece of code as a workaround 

```python
    def fix_embed_globals(N=0):
        # Get the parent frame
        import inspect
        frame_level0 = inspect.currentframe()
        frame_cur = frame_level0
        N = 2  # I may have this value off by one. I rewrote this function to use only standard calls
        for _ix in range(N + 1):
            frame_next = frame_cur.f_back
            if frame_next is None:
                if strict:
                    raise AssertionError('Frame level %r is root' % _ix)
                else:
                    break
            frame_cur = frame_next
        parent_frame = frame_cur
        # Add locals to parent globals
        parent_frame.f_locals.update(parent_frame.f_globals)
        parent_frame.f_globals['_didfixipyglobals'] = True
```

I just call that function whenever I get the name error. It puts all locals in your current frame into the global dictionary. Its hacky but it works in most circumstances.

**Comment by naught101** [comment]
As noted at https://github.com/inducer/pudb/issues/103, a temporary workaround while using an embedded shell is: `globals().update(locals())` (after defining the locally global variables, only).

**Comment by quazgar** [comment]
Hmm, can the milestone of this bug be changed to one of the next versions?

**Comment by Carreau** [comment]
Done.

**Comment by dellis23** [comment]
I don't understand, and it makes me angry.

**Comment by takluyver** [comment]
Yup, I think it was fixed for Django 1.6. If you really can't upgrade, you might want to apply the fix manually. It looks like this was it: https://github.com/django/django/commit/3570ff734e93f493e023b912c9a97101f605f7f5

**Comment by liushapku** [comment]
I found a workaround today that is posted at [#10695](https://github.com/ipython/ipython/issues/10695). A simpler treatment for the case where the IPython is not embedded inside a function is shown at [this thread](https://stackoverflow.com/questions/35161324/how-to-make-imports-closures-work-from-ipythons-embed/51036158#51036158). I am not an expert so please help to check the validity.

**Comment by limwz01** [comment]
I have a workaround posted at https://gist.github.com/limwz01/7e06cacf31d4ed83aa74134971030967 which is not as limited as the current approaches, but with several caveats:
* exceptions raised in the shell further trigger a weird exception
`executing.executing.NotOneValueFound` during traceback exception formatting
and this disables pretty tracebacks
* direct changes to local variables are only visible to code within the
shell session
* outer scope variables must already have had a closure referencing it to be
visible (`nonlocal` does not work) and they become local variables (and
hence above restriction applies)
* local variables named `_ipy_magic` will be overwritten
* return statements without an enclosing function does not give an error and
results in unpredictable behaviour

**Comment by Erotemic** [comment]
I'm wondering if there is any change in Python or IPython that could provide a first-class work around for this issue. It even confuses and annoys @3b1b: https://youtu.be/rbu7Zu5X1zI?si=3gcC1r2k3WrfZu-I&t=1734

**Comment by limwz01** [comment]
My workaround at https://github.com/limwz01/ipython_utils is clunky but works.. and I've been using it for a long while now.

**Comment by Carreau** [comment]
Yeah, I'm in the process of watching @3b1b video, and I know this is an ipython limitation. I'd love a PR, I unfortunately dont have the time to dive into it. 

My guess is that it's the same issue that affects ipdb where list comprehension don't work.

@limwz01, if you think your utils can be ported to be default in IPython, please send a PR. 
Worse case we'll find some issues and that can be reverted, but at least we'll move forward on our comprehension of what it happening.

**Comment by limwz01** [comment]
@gaogaotiantian recently pushed a similar solution to pdb in cpython itself: https://github.com/python/cpython/pull/111094

His solution is a little too simplistic and did not handle all the edge cases, so it may not be suitable here. However, it shows that such a solution is actually acceptable.

**Comment by gaogaotiantian** [comment]
> His solution is a little too simplistic and did not handle all the edge cases, so it may not be suitable here.

Feel free to submit an issue about the edge cases that `pdb` can't handle. We might not be able to fix all of those because we need to make the pdb code as maintainable as possible (so playing with `ast` might not be an option), which means we need to balance the code complexity and the capability, but it would be good for us to at least know what are the cases that we can't handle.

**Comment by limwz01** [comment]
@gaogaotiantian I use the following to illustrate the problems:
```python3
#!/usr/bin/env python3.13
g0 = 0
def main():
    import pdb
    pdb.Pdb().set_trace()
if __name__ == "__main__":
    main()
```
The following are the problems:
* not re-using the same Cells for each local (this is really the biggest issue)
```
(Pdb) t = 1
(Pdb) f = lambda: t
(Pdb) t = 2
(Pdb) f()
1
```
* not obtaining the last expression (a pdb test suggests otherwise, but it is actually wrong)
```
(Pdb) t=10; (lambda: t)()
<no output>
```
(it does return a value for `global g; g=1; g`. but I'm at a loss as to where that value comes from)
* giving a weird result with shadowed variables (actually because it bails out):
```
(Pdb) g0 = 1
(Pdb) global g0; t=g0; t=(lambda: t)()
*** NameError: name 't' is not defined
```
* not removing annotations
```
(Pdb) t:int = 1; (lambda: t)()
*** NameError: name 't' is not defined
* using indentation to set up block scope
```
(Pdb) print((lambda:len("""
...   """))())
7
```
* not hoisting globals (this is arguable):
```
(Pdb) global g; g=1
(Pdb) t=2
(Pdb) g=(lambda: t)()
(Pdb) global g; g
1
```

**Comment by limwz01** [comment]
I can't seem to edit my post to correct some typos and cross-reference the issue, but anyway the issue has been submitted here: https://github.com/python/cpython/issues/126958

**Comment by ur001** [comment]
Use `user_ns`. It may be some dict, or `locals() | globals()`, for example.

```python
from IPython import start_ipython
start_ipython(user_ns={})
```
or 
```python
from IPython.terminal.embed import InteractiveShellEmbed
InteractiveShellEmbed(user_ns={})()
```
or 
```python
import IPython
IPython.embed(user_ns={})
```

Result:
```python
In [1]: a = 1

In [2]: def f(x):
   ...:     return a * x
   ...:

In [3]: f(2)
Out[3]: 2
```

**Comment by limwz01** [comment]
This issue is about the problems of embedding a shell in a function scope,
something you totally missed in your examples, @ur001

**Comment by krassowski** [comment]
@limwz01 it appears that you problem was a different issue, related to pdb, fixed in Python 3.13 and backported to IPython in https://github.com/ipython/ipython/pull/14933.

Passing `locals()` scope explicitly is now required in Python 3.13 as per:
- https://github.com/python/cpython/issues/118888
- https://github.com/python/cpython/pull/119893

It looks like we could just ensure that embed defaults to pass a `locals()` in. Is there any example where 

```python
import IPython
IPython.embed(user_ns={})
```


or 

```python
import IPython
IPython.embed(user_ns=locals())
```

would not work?

**Comment by limwz01** [comment]
Yes, definitely. What you are doing is running the shell with its globals set to the `locals()` at that point in question. That is bad because even global variables will now be "out of scope" in the new shell. Here's a piece of code that has all the types of variables:

```python
import IPython
x0 = 100
def main():
    y0 = 200
    y1 = 300
    def f(z0, z1):
        nonlocal y1
        def g():
            nonlocal z1
            z1 += 1
        y1 += 1
        print(x0, y0, y1, z0, z1)
        IPython.embed(user_ns=locals())
        print(x0, y0, y1, z0, z1)
        IPython.embed(user_ns=locals())
    f(400, 500)
main()
```

Now, let's see what you can do in the first shell:
1. `x0` cannot be accessed
2. no changes to the local variables are retained, i.e. changing any of the variables in the first shell wouldn't have any effect on the second `print`, or their values in the second shell.
3. `g(); print(z1)` changes only the real `z1` and the shell's `z1` is unchanged

Quoting from my readme at https://github.com/limwz01/ipython_utils?tab=readme-ov-file#limited-functionality-in-default-ipython-embedded-shell :

> There are a few workarounds for this, when embedding into the scope of a function `f`:
> 1. Use a mocked-up `globals()` including what was in `locals()`
> 2. Use the real `globals()` and copy over `locals()` into the real `globals()`
> 
> Both of the above require us to patch all the `nonlocal` statements referring to locals in the scope of `f` (and parent functions) to become `global`. In the first workaround, we use IPython with a `globals()` namespace that have been updated with both the original frame's `globals()` and `locals()`, either by running `IPython.start_ipython` with `user_ns` as the mocked-up `globals()` or patching `IPython.terminal.embed.InteractiveShellEmbed` (instead of this, one could possibly keep calling `globals().update(locals())` for every statement executed but that is really inefficient and only works if the created child functions do not modify the referenced variables). In this workaround, references to real globals in created functions will no longer access/modify the real global variable, but a copy, hence diverging from functions created by other means which use the real globals. In the second workaround, the real global namespace gets polluted and globals might have their values incorrectly overwritten by the locals, causing some existing functions to behave incorrectly.

Instead, one could use the (almost) complete solution that I developed (after `pip install ipython_utils`), and everything just works fine, you can even create a closure that works in the second shell:

```python
import ipython_utils
x0 = 100
def main():
    y0 = 200
    y1 = 300
    def f(z0, z1):
        nonlocal y1
        def g():
            nonlocal z1
            z1 += 1
        y1 += 1
        print(x0, y0, y1, z0, z1)
        ipython_utils.embed([f, lambda: z1])
        print(x0, y0, y1, z0, z1)
        ipython_utils.embed([f, lambda: z1])
    f(400, 500)
main()
```

In the first shell:
```python
def h():
    nonlocal z1
    z1 += 2
```

In the second shell:
```python
print(z1)
g()
print(z1)
h()
print(z1)
```

And this prints:
```
500
501
503
```

If anyone can break my code, feel free to post an issue! EDIT: I've updated my code due to a [backward-incompatible change in Python3.13](https://github.com/python/cpython/issues/136308).
